---
title: "Integrating RNA and ATAC through WNN"
author: "kou"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,include=T,message=F,eval=F, warning=F)
```

```{r libraries, warning=F}
library(Signac)
library(Seurat)
library(tidyverse)
library(ggpubr)
library(shiny)
library(qs)
library(reticulate)
library(clustree)
library(ArchR)
library(reticulate)
library(MotifDb)
library(universalmotif)
library(BSgenome.Hsapiens.UCSC.hg38)
library(EnsDb.Hsapiens.v86)
library(TFBSTools)
library(BiocParallel)

# always set seed!!!
set.seed(105)

options(future.globals.maxSize = 4000 * 1024^2)
addArchRThreads(threads = 125, force = T)
addArchRGenome("hg38")

# set python
use_python("/g/scb/zaugg/kou/projects/composite/analysis/sc_py_env/bin")
use_virtualenv("/g/scb/zaugg/kou/projects/composite/analysis/sc_py_env")
```

```{r data_import}
ln_id = c("LN0025","LN0177","LN0193","LN0438")
atac_integrated = qread("02_integration/rds/02_atac_integrated.qs")
sct_integrated = qread("02_integration/rds/03_sct_integrated.qs")
```

### Add modalities together
```{r data}
# set default assays
DefaultAssay(sct_integrated) = "integrated"
DefaultAssay(atac_integrated) = "ATAC"

# add assays to sct
sct_integrated[["ATAC"]] = atac_integrated[["ATAC"]]
sct_integrated[["GeneScoreImputed"]] = atac_integrated[["GeneScoreImputed"]]

# transfer embeddings
sct_integrated[["atacUMAP"]] = CreateDimReducObject(embeddings = Embeddings(atac_integrated@reductions$atacUMAP),
                                                    assay = "ATAC",
                                                    key = "atacUMAP_")

sct_integrated[["integrated_lsi"]] = CreateDimReducObject(embeddings = Embeddings(atac_integrated@reductions$integrated_lsi),
                                                          assay = "ATAC",
                                                          key = "integratedlsi_")
```

### WNN analysis
- Notes on clustering (recommendation to do in PCA space, not UMAP: https://github.com/satijalab/seurat/issues/1081)
- Set k.nn = 30 to match n.neighbors in `RunUMAP`

TO DO: benchmark with different k.nn, default was not enough
```{r wnn}
wnn = FindMultiModalNeighbors(sct_integrated,
                              reduction.list = list("sctPCA","integrated_lsi"),
                              dims.list = list(1:50,2:50),
                              k.nn = 30)

# create wnn embeddings and perform clustering
wnn = RunUMAP(wnn, 
              nn.name = "weighted.nn", 
              reduction.name = "wnnUMAP", 
              reduction.key = "wnnUMAP_",
              n.neighbors = 30L)

# also create 3D UMAP
wnn = RunUMAP(wnn, 
              nn.name = "weighted.nn", 
              reduction.name = "wnnUMAP.3D", 
              reduction.key = "wnnUMAP.3D_",
              n.neighbors = 30L,
              n.components = 3)

# find optimal cluster with clustree
wnn = FindClusters(wnn,
                   resolution = seq(0.8, 1.6, by = 0.4),
                   graph.name = "wsnn",
                   algorithm = "4",
                   method = "igraph",
                   random.seed = 1,
                   group.singletons = F,
                   verbose = F)

qsave(wnn,"02_integration/rds/04_wnn.qs",nthreads=16)
```


Visualizing clustering results
```{r clustree}
clustreeplot = clustree(wnn@meta.data, prefix = "wsnn_res.")
ggsave("02_integration/figs/04_clustree_wnn.png", clustreeplot, height = 8, width = 10)
clustreeplot
```

Cluster behavior in UMAP
```{r umapClusters}
wnnUMAPembed = as_tibble(Embeddings(wnn,"wnnUMAP"), rownames = "cell_id")

wnnUMAPembed$sample = wnn$orig.ident

# add cluster info
wnnUMAPembed = cbind(wnnUMAPembed, wnn@meta.data[,paste0("wsnn_res.",seq(0.8, 1.6, by = 0.4))])

# pivot longer
clusterVis = pivot_longer(wnnUMAPembed, 
                          cols = paste0("wsnn_res.",c("0.8","1.2","1.6")),
                          names_to = "res",
                          values_to = "cluster_id")

wnnUMAPembed %>% 
  subset(., wsnn_res.1.6 == "21") %>%
  ggplot(., aes(wnnUMAP_1,wnnUMAP_2))+
  geom_jitter(aes(color = wsnn_res.1.2), 
              show.legend = T, 
              size = 0.1)+
  theme_bw() +
  theme(aspect.ratio = 3/3) +
  facet_grid(cols = vars(sample)) +
  labs(title = "wsnn_res.1.6 == 21")
```
It seems like wsnn_res.1.6 is overclustered (i.e. cluster 21 from res 1.6 actually is 2 separate clusters in res 1.2(1 and 22)).
Furthermore, the additional cluster from 1.6 (1 and 21) 
For now use 1.2.

```{r setIdents}
Idents(wnn) = "wsnn_res.1.2"

# setting idents does not reset seurat clusters
wnn$seurat_clusters = Idents(wnn)

# remove singletons
wnn = subset(wnn, idents = "singleton", invert = T)

# save
qsave(wnn,"02_integration/rds/04_wnn.qs", nthreads=16)
```

### Visualization
UMAP: All samples in same embedding
```{r umap}
# plot
wnnUMAP_allSamples = ggplot(wnnUMAPembed, aes(wnnUMAP_1,wnnUMAP_2)) +
  geom_jitter(aes(color = sample), size = 0.15, show.legend = T) +
  theme_bw() +
  theme(aspect.ratio = 3/3,
        panel.grid = element_blank(),
        strip.background = element_blank(),
        strip.text = element_blank())

ggsave("02_integration/figs/04_wnnUMAP_allSamples.png", wnnUMAP_allSamples, width = 5, height = 5)

wnnUMAP_allSamples
```

UMAP: Clusters (wnn_res.1.2) faceted by samples. 
```{r umapClusters}
# plot
wnnUMAP_clusters = ggplot(wnnUMAPembed, aes(wnnUMAP_1,wnnUMAP_2)) +
  geom_jitter(aes(color = wsnn_res.1.2), size = 0.05, show.legend = T) +
  theme_bw() +
  theme(aspect.ratio = 3/3,
        panel.grid = element_blank(),
        strip.background = element_blank()) +
  facet_grid(cols = vars(sample))

ggsave("02_integration/figs/04_wnnUMAP_clusters.png", wnnUMAP_clusters, width = 10, height = 5)

wnnUMAP_clusters
```
Subset RNA features to color code embeddings
```{r features}
# genes of interest
genes = c("MS4A1",
          "PAX5",
          "CXCR5",
          "CD3D",
          "CD4",
          "CD8A",
          "FOXP3",
          "TBX21",
          "GATA3",
          "BCL6",
          "PDCD1",
          "HAVCR2",
          "BCL2",
          "MKI67")

# get norm counts
normcounts = wnn@assays$RNA@data
normcounts = normcounts[genes,]
```

UMAP: Facet by samples, color by normalized RNA counts.
TO DO: recolor so alpha = 0.5 if count is 0
```{r umapFeature}
# join, after confirming all(rownames(sctUMAPembed) == colnames(normcounts))
wnnUMAPembed = cbind(wnnUMAPembed, t(as.matrix(normcounts)))

# create a function for umap plotting
# TO DO: rescale to remove outliers, or only plot q99, q1
umapFeature = function(df_embed,locus){
  ggplot(df_embed, aes(wnnUMAP_1,wnnUMAP_2)) +
  geom_jitter(aes(color = .data[[locus]]), size = 0.05) +
  scale_color_viridis_c(option = "turbo") +
  theme_bw() +
  theme(panel.grid = element_blank(),
        aspect.ratio = 3/3,
        strip.background = element_blank()) +
  facet_grid(cols = vars(sample))
}

# save, TO DO: lapply
ms4a1 = umapFeature(wnnUMAPembed, "MS4A1")
cd3d = umapFeature(wnnUMAPembed, "CD3D")
bcl6 = umapFeature(wnnUMAPembed, "BCL6")
bcl2 = umapFeature(wnnUMAPembed, "BCL2")

ggsave("02_integration/figs/04_wnnUMAP_MS4A1.png", ms4a1, width = 10, height = 5)
ggsave("02_integration/figs/04_wnnUMAP_CD3D.png", cd3d, width = 10, height = 5)
ggsave("02_integration/figs/04_wnnUMAP_BCL6.png", bcl6, width = 10, height = 5)
ggsave("02_integration/figs/04_wnnUMAP_BCL2.png", bcl2, width = 10, height = 5)
```


### Peak calling on wnn clusters using MACS2
Liked ArchR's iterative peak overlap process on fixed with peaks, which prevents peaks from "daisy-chaining" into larger peaks and masking small peaks.

Need to first create an ArchR project with all samples/cells, and add cluster assignment and wnn embedding.
```{r archRdata}
archR_keep = loadArchRProject("01_QC/archRprojects/ArchRproj_keep/")

# also remove singletons
archR_keep = archR_keep[gsub("#","_",archR_keep$cellNames) %in% Cells(wnn),]

# check if cell order is the same
all(gsub("#","_",archR_keep$cellNames) == Cells(wnn))

# order archR to match seurat
ord = match(Cells(wnn),gsub("#","_",archR_keep$cellNames))
archR_keep = archR_keep[ord,]

all(gsub("#","_",archR_keep$cellNames) == Cells(wnn))

# add cluster info
archR_keep$wsnn_cluster = wnn$wsnn_res.1.2

# add wnn embedding
# checked all(gsub("#","_",archR_keep$cellNames) == rownames(wnnUMAP))
wnnUMAP = Embeddings(wnn, reduction = "wnnUMAP")

# rename rows and cols to be consistent w archr
rownames(wnnUMAP) = gsub("_","#", rownames(wnnUMAP))
colnames(wnnUMAP) = paste0("seurLSI#",colnames(wnnUMAP))

archR_keep@embeddings$wnnUMAP = SimpleList(df = wnnUMAP, params = list())

# save for future loading
allLN = saveArchRProject(ArchRProj = archR_keep,
                         outputDirectory ="02_integration/archRprojects/allLN_ArchR", 
                         load = T)

#allLN = loadArchRProject("02_integration/archRprojects/allLN_ArchR")
```

Call peaks using macs2 on pseudobulked replicates of wnn clusters identified in Seurat.
```{r callPeaks}
h5disableFileLocking()

# get pseudobulk clusters
allLN = addGroupCoverages(
  ArchRProj = allLN, 
  groupBy = "wsnn_cluster", 
  threads = 125,
  force = T
)

h5enableFileLocking()

# call peaks
allLN = addReproduciblePeakSet(
  ArchRProj = allLN,
  minCells = 50,
  groupBy = "wsnn_cluster", 
  pathToMacs2 = "sc_py_env/bin/macs2",
  excludeChr = c("chrM","chrY"),
  threads = 1,
  force = T
)

# add peak matrix
allLN = addPeakMatrix(allLN, 
                      force = T)

# save project
allLN = saveArchRProject(ArchRProj = allLN,
                         outputDirectory ="02_integration/archRprojects/allLN_ArchR", 
                         load = T)
```

### Motif annotation
TO DO: add explanation
```{r motifdb}
# get PFM matrices
jaspar2022=query(MotifDb, andStrings=c("hsapiens","jaspar2022"))
hocomocoV11=query(MotifDb, andStrings=c("hsapiens","HOCOMOCOv11"))

# can access additional information w/ mcols
jaspar_meta = mcols(jaspar2022)
hoco_meta = mcols(hocomocoV11)

# check to see that PAX5, BCL6, and other TFs are in the databases
tfsOfI = c("PAX5","TCF7","BCL6","IRF8","IRF4","FOX")

tfsInJaspar = names(jaspar2022)[grep(str_c(tfsOfI, collapse = "|"), names(jaspar2022))]

tfsInHoco = names(hocomocoV11)[grep(str_c(tfsOfI, collapse = "|"), names(hocomocoV11))]

sort(tfsInJaspar)
sort(tfsInHoco)
```

Both seem to cover (super brief list of) TFs in interest. There are differences in the PFM however.\
Example: FOXP3

```{r FOXP3pfm, eval=F}
# example: FOXP3 PFM in jaspar
jaspar2022[[grep("-FOXP3-",names(jaspar2022))]]

# example: FOXP3 PFM in hoco
hocomocoV11[[grep("-FOXP3_",names(hocomocoV11))]]
```
```{r finalMotifList}
# tidy jaspar, want TF-unique/matID_db
## use different separators, easier to separate later for plotting, etc
names(jaspar2022) = gsub("Hsapiens.*jaspar2022-","",names(jaspar2022)) %>% 
  paste0(.,"_jaspar2022")

## change rownames of elementMetadata
rownames(jaspar2022@elementMetadata) = names(jaspar2022)

# tidy hoco
## subset hocomoco w/ qualities A-C only
hoco_ABC = hocomocoV11[!grepl(".D$",names(hocomocoV11))]

## tidy hoco names
names(hoco_ABC) = gsub("Hsapiens.*-[A-C]-|HUMAN.","",names(hoco_ABC)) %>%
  gsub("_","-",.) %>%
  paste0(.,"_hocoV11")

## change rownames of elementMetadata
rownames(hoco_ABC@elementMetadata) = names(hoco_ABC)

# combine list, convert into universalmotif class, save bc metadata is useful
finalMotif = c(jaspar2022,hoco_ABC)
qsave(finalMotif, "02_integration/rds/04_finalMotif.qs", nthreads = 16)
```

Annotate peaks with motifs contained in PWM list.
```{r peakAnno}
# convert to PWM
finalPWM = convert_motifs(finalMotif, class = "TFBSTools-PWMatrix")
finalPWM = do.call(PWMatrixList,finalPWM)
names(finalPWM) = names(finalMotif)

# add motif annotations to peak
allLN = addMotifAnnotations(allLN, 
                            motifPWMs = finalPWM,
                            motifSet = NULL,
                            collection = NULL,
                            annoName = "Motif",
                            force = F)

# add background peaks
allLN = addBgdPeaks(allLN)

# save project
allLN = saveArchRProject(ArchRProj = allLN,
                         outputDirectory ="02_integration/archRprojects/allLN_ArchR", 
                         load = T)
```

## Add peaks to combined wnn Seurat object
```{r createChromatinAssay}
# get SE
peakSE = getMatrixFromProject(allLN,"PeakMatrix")

# get counts in peaks, rename to match seurat obj
peakMat = assay(peakSE)
colnames(peakMat) = gsub("#","_",colnames(peakMat))

# get ranges of peaks; idx is index of chr
peakRanges = granges(peakSE)

# get metadata to add to seurat metadata (ie: Frac Reads in Peaks)
data = colData(peakSE)
rownames(data) = gsub("#","_",rownames(data))

# get annotation
hg38_anno = ArchRtoSignac::getAnnotation(reference = EnsDb.Hsapiens.v86, 
                                         refversion = "hg38",
                                         seqStyle = "UCSC")

# create chromatin assay
peakAssay = CreateChromatinAssay(counts = peakMat,
                                 ranges = peakRanges,
                                 genome = "hg38",
                                 min.cells = 0,
                                 annotation = hg38_anno,
                                 validate.fragments = F)

# add "peaks" assay to seurat obj (TO DO: rename other assay to "tile"?)
wnn[["Peaks"]] = peakAssay

# reorder
## first checked that all(rownames(wnn@meta.data) == Cells(wnn))
ord = match(Cells(wnn), rownames(data))
data = data[ord,]

# add to metadata
wnn@meta.data = cbind(wnn@meta.data, data)
```

Adding motif data to peaks assay
```{r addMotifObject}
# get peak annotation info from ArchR project
peakAnno = getPeakAnnotation(allLN, "Motif")

# get granges list of motif positions
positions = readRDS(peakAnno$Positions)

# get motif info in peaks
matchesSE = readRDS(peakAnno$Matches)

# get binarized matrix of peaks (rows) vs motif (cols)
motifMat = assay(matchesSE)

# rename rownames
rownames(motifMat) = GRangesToString(matchesSE@rowRanges)

# get granges containing metadata of peakset; previously checked if same as peakRanges
peakMetadata = as.data.frame(matchesSE@rowRanges, row.names = GRangesToString(matchesSE@rowRanges))

# create motif object
motifObj = CreateMotifObject(data = motifMat,
                             pwm = allLN@peakAnnotation$Motif$motifs,
                             motif.names = names(allLN@peakAnnotation$Motif$motifs),
                             positions = positions,
                             meta.data = peakMetadata)

# add motif object to peaks seurat object
DefaultAssay(wnn) = "Peaks"
Motifs(wnn) = motifObj

# save and analyze in separate script
qsave(wnn,"02_integration/rds/04_wnn.qs", nthreads = 64)
```

## Identify B and T cell clusters
Clusters (defined by wsnn_res.0.8), can be predicted to be B cells based on MS4A1 vs CD3D expression 

TO DO: a more automated way with add module score?
```{r violin}
# relevel
wnn@active.ident = factor(wnn@active.ident, levels = 1:23)

# visualize ms4a1 cluster
DefaultAssay(wnn) = "RNA"

# idents set as wsnn_res.0.8 as default since it was most recent cluster
VlnPlot(wnn,
        features = c("LYZ","SLC8A1"),
        assay = "RNA",
        slot = "data")
```

Based on log-normalized RNA ("RNA" assay, "data" slot):
- T cells (CD3D, CD3E): 1,4,5,8,9,10,11,12,13,15,16,19
- B cells (MS4A1, PAX5): 2,3,6,7,14,17,18,20
- pDC (CD4high, IL3RA, CLEC4C): 21
- NK (KLRD1, NCAM1): 22
- myeloid (LYZ,CST3,SLC8A1): 23

These results also aligned with "GeneScoreImputed" assay. 

Add preliminary annotation
```{r prelimAnno}
Tclusters = c("1","4","5","8","9","10","11","12","13","15","16","19")
Bclusters = c("2","3","6","7","14","17","18","20")

prelim = case_when(wnn@active.ident == "21" ~ "pDC",
                   wnn@active.ident == "22" ~ "NK",
                   wnn@active.ident == "23" ~ "myeloid",
                   wnn@active.ident %in% Tclusters ~ "T", 
                   wnn@active.ident %in% Bclusters ~ "B")

# add to wnn object
wnn$anno_prelim = factor(prelim, levels = c("B","T","NK","pDC","myeloid"))
```

### Visualization
```{r dimplot}
DimPlot(wnn, reduction = "wnnUMAP", group.by = "anno_prelim", split.by = "orig.ident")
```
```{r finalwnnSave}
qsave(wnn, "02_integration/rds/04_wnn.qs", nthreads = 32)
```



