---
title: "ATAC peak calling"
author: "kou"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r import, warning=F, message=F}
library(Seurat)
library(Signac)
library(ArchR)
library(BSgenome.Hsapiens.UCSC.hg38)
library(parallel) # need to set threads > 1
library(presto)
library(tidyverse) 
library(ggpubr)
library(shiny)

theme_set(theme_bw())

set.seed(105)
addArchRThreads(threads = 32, force = T)
options(future.globals.maxSize = 4000 * 1024^2)

# this step includes ref genome, TxDb object, and blacklist!
addArchRGenome("hg38")
```

## Creating Chromatin Assay for integration
In order to perform data integration with WNN in Seurat, need to create a `Chromatin Assay` object to add to Seurat object. 
Chromatin Assay needs:
1. raw counts OR data (normalized counts); cannot supply both! features (rows) x cells (cols)
2. GRanges corresponding to rows
3. motif object > need to check what format this is supposed to be
4. fragments > either path, or a `Fragments Object` > check to see what this looks like
5. genome
6. annotation for genome 
7. bias for Tn5 insertion (optional??)
8. additional named matrices of enrichment in specific regions (ie: TSS or ChIP-seq data)

```{r import data, include=T}
# load archR proj after filtering cells from RNA and ATAC
archR_keep = loadArchRProject("01_QC/ArchRproj_keep/")

# separate archR by samples
archR_LN0025 = archR_keep[archR_keep$Sample == "LN0025",]
archR_LN0177 = archR_keep[archR_keep$Sample == "LN0177",]
archR_LN0193 = archR_keep[archR_keep$Sample == "LN0193",]
archR_LN0438 = archR_keep[archR_keep$Sample == "LN0438",]

# create list of projects to process individually
archR_all = c(archR_LN0025,archR_LN0177,archR_LN0193,archR_LN0438)

# create list of outputDirectory to save ArchR projects
outDir = paste0("01_QC/ArchR_",c("LN0025","LN0177","LN0193","LN0438"))

# save all projects
map2(archR_all, outDir, ~ saveArchRProject(ArchRProj = .x, outputDirectory = .y, load = F))

# test loading projects
archR_all = lapply(outDir,loadArchRProject)
```


### LSI reduction
Perform LSI clustering using 500bp genomic tiles (not peaks like in Signac).
Test how different number of iterations affect clustering! Add "LSI_2" vs "LSI_4"
Will increase resolution for each iteration (should be iterations - 1).
The resolution here is NOT for assigning cluster identity, it is to help identify "average accessibility" for each cluster for linear reduction.

TO DO: Add additional LSI!!! LSI_6 for sure
```{r lsi, include=T}
# can also have multiple resolutions for clustering, use the same number as in gex norm
# clusterParams are same as in Seurat
# test 2 vs 4 iterations to see if noticeable difference
archR_all = map(archR_all, ~ addIterativeLSI(.x,
                                             useMatrix = "TileMatrix",
                                             name = "LSI_2", 
                                             iterations = 2,
                                             clusterParams = list(resolution = 0.8,
                                                                  sampleCells = 10000,
                                                                  n.start = 10), 
                                             varFeatures = 25000, 
                                             dimsToUse = 1:30))

archR_all = map(archR_all, ~ addIterativeLSI(.x,
                                             useMatrix = "TileMatrix",
                                             name = "LSI_4", 
                                             iterations = 4, 
                                             clusterParams = list(resolution = c(0.6,0.8,1.0),
                                                                  sampleCells = 10000,
                                                                  n.start = 10), 
                                             varFeatures = 25000,
                                             dimsToUse = 1:30))

# save projects
map2(archR_all, outDir, ~ saveArchRProject(ArchRProj = .x, outputDirectory = .y, load = F))

# load projects
archR_all = lapply(outDir,loadArchRProject)
```

### Clustering
Perform clustering on both LSI_2 and LSI_4 dimensional reductions, compare output. Start with resolution 1.15.
TO DO: need to nest map2 within lapply to test different resolutions
TO DO: try with leiden clustering
```{r clustering, include=T}
## clustering on LSI_2 first
archR_all = map(archR_all, ~ addClusters(.x,
                                         reducedDims = "LSI_2",
                                         method = "Seurat",
                                         name = "LSI2res1.15",
                                         resolution = 1.15,
                                         dimsToUse = 1:30))

## clustering on LSI_4
archR_all = map(archR_all, ~ addClusters(.x,
                                         reducedDims = "LSI_4",
                                         method = "Seurat",
                                         name = "LSI4res1.15",
                                         resolution = 1.15,
                                         dimsToUse = 1:30,
                                         force = T))

# save projects
map2(archR_all, outDir, ~ saveArchRProject(ArchRProj = .x, outputDirectory = .y, load = F))

archR_all = lapply(outDir,loadArchRProject)
```

### UMAP embeddings 
Add umap based on LSI4
```{r umap, include=T}
archR_all_umap = map(archR_all, ~ addUMAP(.x,
                                     reducedDims = "LSI_4",
                                     name = "atacUMAP",
                                     seed = 105,
                                     #dimsToUse = 1:30,
                                     outlierQuantile = 0.99,
                                     force = T))

# quick plot
lapply(archR_all_umap, plotEmbedding, colorBy = "cellColData", name = "LSI4res1.15", embedding = "atacUMAP")
```

### Get imputed weights to be used for gene score activity matrix
Use LSI 4 dimen reduction
```{r impute weights, include = T}
archR_all_umap = lapply(archR_all_umap, addImputeWeights,reducedDims = "LSI_4",dimsToUse=1:30,sampleCells=10000)

# save projects
map2(archR_all_umap, outDir, ~ saveArchRProject(ArchRProj = .x, outputDirectory = .y, load = F))
archR_all_umap = lapply(outDir,loadArchRProject)
```

### Plot UMAP embedding 
Want to get a sense how heterogeneous the B and T cell populations are. 
```{r umap reformat before plot, include = T}
# get unimputed matrix, which is contained in the summarized experiment object
# gene names are accessed in elementMetadata and are all the same!
# TO DO: can remove genes that are 0 for all cells
geneScoreMat = lapply(archR_all_umap, getMatrixFromProject, useMatrix = "GeneScoreMatrix", verbose=F)
geneNames = geneScoreMat[[1]]@elementMetadata$name

# get impute weights
imputeWeights = lapply(archR_all_umap, getImputeWeights)

# get matrix w impute scores, need to use se object in geneScoreMat to get gene names (rows)
imputedMat = map2(geneScoreMat,imputeWeights, ~ imputeMatrix(assay(.x),.y))

# reformat: add gene names, transpose
imputedMat = lapply(imputedMat, function(x){
  rownames(x) = geneNames
  x = t(x)
  return(x)
})

# remove columns with colsum = 0 (not expressed in any cell)
imputedMat = lapply(imputedMat, function(x){
  # get which cols have colsum != 0 (gene expressed in some cells)
  whichToKeep = colSums(x) != 0
  
  # subset mat w/ cols to keep
  sub = x[,whichToKeep]
  
  # round for readability
  sub = round(sub, digits = 4)
  return(sub)
})

# get umap embedding
umapEmbed = lapply(archR_all_umap, getEmbedding, embedding = "atacUMAP")

# rename umap embedding columns, add sample name 
umapEmbed = map(umapEmbed, ~ as_tibble(.x, rownames = "cellID") %>%
                  rename("atacUMAP_1" = 2, "atacUMAP_2" = 3) %>%
                  mutate(Sample = str_replace(cellID,"#.*","")))

# add clustering info from colData of ArchR project
umapEmbed = map2(umapEmbed, archR_all_umap, ~ .x %>%
                   mutate(cluster = .y$LSI4res1.15))

# cbind imputed matrix w/ genes, checked that rownames of imputedMat == umapEmbed$cellID
# map2(imputedMat, umapEmbed, ~all(rownames(.x)==.y$cellID))
umapPlot = map2(umapEmbed, imputedMat, ~ cbind(.x,.y))

# save rds
saveRDS(umapPlot,"01_QC/rds/05_umapPlot.rds")
umapPlot = readRDS("01_QC/rds/05_umapPlot.rds")
```

```{r}
# test ggplot before shiny app
var = "MS4A1"

lapply(umapPlot, function(x){
  ggplot(x, aes(atacUMAP_1,atacUMAP_2))+
  geom_jitter(size = 0.25, aes(color = .data[[var]])) +
  scale_color_viridis_c(option = "plasma") +
  theme(aspect.ratio = 3/3,
        panel.grid = element_blank())
})
```

TO DO: fix shiny app, needs to be individual or need to create cow_plot
```{r umap shiny, include=T}
ui = fluidPage(
  titlePanel("UMAP of scATAC-seq features, colored by imputed gene score activity"),
  sidebarLayout(
    
    sidebarPanel(
      
      # type gene of interest
      textInput(inputId = "SYMBOL",
                label = "type symbol of interest: ",
                value = "PAX5"),
    ),
    
    mainPanel(plotOutput(outputId = "UMAP_grids"))
  )
)

server = function(input,output){
  # ggplot fxn to lapply
  plotUMAP = function(df){
    ggplot(df, aes(atacUMAP_1,atacUMAP_2))+
      geom_jitter(size = 0.25, aes(color = .data[[input$SYMBOL]])) +
      scale_color_viridis_c(option = "plasma") +
      theme(aspect.ratio = 3/3,
      panel.grid = element_blank())
  }
  
   output$UMAP_grids = renderPlot({
    lapply(umapPlot,plotUMAP)})
}



```

